### PR TITLE
Add Automatic Mixed Precision (AMP) support for derivatives.

### PR DESCRIPTION
**Pull request by Alexey-Kamenev**
# Modulus Pull Request

## Description

Training Physics Informed Neural Networks with Automatic Mixed Precision (AMP) currently leads to infinity loss for several models. This comes from the higher-order derivatives, which could go beyond the FP16 dynamic range (`5.96e-8` to `65504`). For example, the training of lid driven cavity problem got terminated around 3000 steps because `u__y__y` at the corner passes the FP16 maximum value of `65504`. 

The default AMP [GradScaler](https://pytorch.org/tutorials/recipes/recipes/amp_recipe.html#adding-gradscaler) tracks the model parameter gradients but not the derivatives calculated from `torch.autograd.grad`.

## DerivScaler

The derivatives need to be tracked by another scaler because the derivatives and NN parameter gradients have different dynamic ranges.
The dynamic range of FP16 is from 2^-24 to 2^15 (40 powers of 2).  
The following range is different from problem to problem, just for reference
- Typical weight gradient range: 2^-40 to 2^-10
- Typical first order derivative range: 2^-10 to 2^5
- Typical second order derivative range: 2^0 to 2^20

This PR adds a DerivScaler which
1. scales and unscales the derivatives during forward in the derivative node, so the operations used in FP16 are in the good range
2. check INFs/NaNs when unscaling the derivatives. When INFs/NaNs are detected, this iteration will be skipped, and the scale value will be adjusted.

It supports the following features
- Per derivative order scalers (default)
- Per derivative term scalers
- Control the scaling factors
  - Avoid a very low derivative scaling factor
    - use fused activation or disable autocast for activation to avoid intermediate result overflow.
    - Use FP32 for the first layer that produces the high-order derivatives, this layer always overflows and this does not introduce too much perf drop.
  - Avoid scaling factor decreasing too fast, useful for Fourier Neural Network
    - Entering “recover mode” if the scaling factor is less than a predefined threshold. In this mode, the scaling factor grows more frequently.

For more details, please refer to [this publication](https://ml4physicalsciences.github.io/2022/files/NeurIPS_ML4PS_2022_4.pdf).
## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus-sym/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus-sym/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus-sym/issues) is linked to this pull request.

### CODE REVIEW

File: modulus/sym/amp.py [hunk 1]
```diff
@@ -0,0 +1,634 @@
...
+        max_scale=2.0**0,
+        recover_threshold=2.0**-6,
+        recover_growth_interval=100,
+        **kwargs,
+    ):
+        """
+        DerivScaler based on the GradScaler.
+
+        The main difference between DerivScaler and GradScaler is that GradScaler unscales
+        the optimizer's weight gradients, while DerivScaler unscales the derivative tensors.
```
Comment (jinzex) [inline modulus/sym/amp.py:352]:
```suggestion
        The main difference between DerivScaler and GradScaler is that GradScaler unscales 
        the weight gradients (maintained by the optimizer), while DerivScaler unscales the
        data gradients (higher order derivatives).
```

File: modulus/sym/amp.py [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+        """
+        Updates the scale factor.
+
+        If self._found_inf > 0, the scale is multiplied by ``backoff_factor`` to reduce it.
+        If ``growth_interval`` unskipped iterations occurred consecutively, the scale is
+        multiplied by ``growth_factor`` to increase it.
+        If the scaling factor is less or equal to the recover_threshold, it will grow more
+        frequently (using the recover_growth_interval) to avoid decreasing too small and
+        stopping the training process.
+        The max_scale defins a upper bound for the scaling factor.
```
Comment (jinzex) [inline modulus/sym/amp.py:479]:
```suggestion
        If self._found_inf > 0, the scale is multiplied by ``backoff_factor`` to decrease it.
        If there are ``growth_interval`` consecutive unskipped iterations, the scale is
        multiplied by ``growth_factor`` to increase it.
        If the scaling factor is less or equal to the recover_threshold, it will grow more
        frequently (using the recover_growth_interval) to avoid decreasing too small and
        stopping the training process.
        The max_scale defines an upper bound for the scaling factor.
```

File: modulus/sym/domain/domain.py [hunk 3]
```diff
@@ -291,6 +292,20 @@ def add_ntk(self, ntk: NTK):
         self.ntk = ntk
         self.ntk_weights = {}
 
+    def setup_deriv_scaler(self, deriv_scalers: DerivScalers):
+        """
+        Setup derivative scalers.
+
+        Iterate over all constraints and their graphs to setup derivative scalers.
```
Comment (jinzex) [inline modulus/sym/domain/domain.py:299]:
```suggestion
        Iterate over all constraints and their corresponding graphs to set up derivative scalers.
```

File: modulus/sym/eq/derivatives.py [hunk 4]
```diff
@@ -93,20 +97,46 @@ def dict_output(
...
+                var = self._scale(var, var_name)
+
+            # Calculate gradient using Autodiff.
+            with torch.cuda.amp.autocast(enabled=False):
+                # Autograd within the autocast region is not recommended and could
+                # have unpredicted behavior. Autocast has thread-local configurations.
+                # Because Autograd runs in a different thread, so it could have
+                # different/default AMP dtype from the one we specified in the forward
+                # thread. Disabling autocast here will allow Autograd to run the same
+                # dtype as the forward pass.
```
Comment (jinzex) [inline modulus/sym/eq/derivatives.py:126]:
```suggestion
                    # Autocast needs to be disabled because using Autograd within the autocast region
                    # is not recommended and could have unpredictable behavior. Autocast relies on
                    # thread-local configurations, and since Autograd runs in a different thread, it
                    # could use a different/default AMP dtype than the one specified in the forward thread.
                    # Disabling autocast here will allow Autograd to run the same dtype as the forward pass.
```

**Review by jinzex — COMMENTED** [review]
LGTM. 

Considering that FP16 (E5M10) has the same dynamic range as FP8 (E5M2), and the [FP8 recipe](https://arxiv.org/abs/2209.05433) has moved to per-tensor scaling instead of per-global loss-scaling, if we are interested in a more robust recipe,  future directions might include:
1. Recommend `per_term_scaler` instead of `per_order_scaler` by default.
2. Per-tensor scaling (FP16) so that we can use the full dynamic range of FP16 for each forward and backward GEMMs. This will require FP16 GEMMs that can apply scaling factors for input tensors A and B and will introduce some overhead.
3. Explore [FP8 recipe](https://github.com/pytorch-labs/float8_experimental) with dynamic scaling.

File: modulus/sym/amp.py (hunk 1 above)
Comment (Alexey-Kamenev) [reply modulus/sym/amp.py]: Done.

File: modulus/sym/amp.py (hunk 2 above)
Comment (Alexey-Kamenev) [reply modulus/sym/amp.py]: Done.

File: modulus/sym/domain/domain.py (hunk 3 above)
Comment (Alexey-Kamenev) [reply modulus/sym/domain/domain.py]: Done.

File: modulus/sym/eq/derivatives.py (hunk 4 above)
Comment (Alexey-Kamenev) [reply modulus/sym/eq/derivatives.py]: Done.

**Comment by Alexey-Kamenev** [comment]
Thank you for the review, @jinzex! I'll add your future directions suggestions to my list of items to look at for the next release.

**Review by ktangsali — APPROVED** [review]

**Review by ktangsali — APPROVED** [review]
Looks great! We should probably add some docs about the usage of this in a separate PR.
